Python/optimizer.c

```diff
@@ -244,6 +244,7 @@ _Py_ClearExecutorDeletionList(PyInterpreterState *interp)
         ts = PyThreadState_Next(ts);
         HEAD_UNLOCK(runtime);
     }
+    EXECUTOR_LIST_LOCK(interp);
```

Comment:
You do need a lock, but since there's a Py_DECREF inside, you should add them all to a PyListObject/PyTupleObject, then free the executors in the PyListObject after the lock.